valhalla/thor/dijkstras.h

```diff
@@ -0,0 +1,273 @@
+    return RouteCallbackRecommendedAction::no_action;
+  };
+
+  bool has_date_time_;
+  int start_tz_index_;   // Timezone at the start of the Dijkstras
+  sif::TravelMode mode_; // Current travel mode
+  uint32_t access_mode_; // Access mode used by the costing method
+
+  // For multimodal
+  bool date_set_;
+  bool date_before_tile_;
+  uint32_t date_;
+  uint32_t dow_;
+  uint32_t day_;
+  uint32_t max_transfer_distance_;
+  std::string origin_date_time_;
+  uint32_t start_time_;
+  std::unordered_map<std::string, uint32_t> operators_;
+  std::unordered_set<uint32_t> processed_tiles_;
+
+  // Current costing mode
+  std::shared_ptr<sif::DynamicCost> costing_;
+
+  // Vector of edge labels (requires access by index).
+  std::vector<sif::BDEdgeLabel> bdedgelabels_;
+  std::vector<sif::MMEdgeLabel> mmedgelabels_;
+
+  // Adjacency list - approximate double bucket sort
+  std::shared_ptr<baldr::DoubleBucketQueue> adjacencylist_;
+
+  // Edge status. Mark edges that are in adjacency list or settled.
+  EdgeStatus edgestatus_;
+
+  /**
+   * Initialize prior to computing the Dijkstrass. Creates adjacency list,
+   * edgestatus support, and reserves edgelabels.
+   * @param bucketsize  Adjacency list bucket size.
+   */
+  template <typename label_container_t>
+  void Initialize(label_container_t& labels, const uint32_t bucketsize);
+
+  /**
+   * Sets the start time for forward expansion or end time for reverse expansion based on the
+   * locations date time string and the edge candidates timezone
+   *
+   * @param location           which location to use for the date time information
+   * @param node_id            the node from which to get timezone information
+   * @param reader             the reader for looking up timezone information
+   * @returns                  a pair with the first being the epoch seconds for the date time at that
+   *                           timezone and the second being the ordinal second from the beginning of
+   *                           the week
+   */
+  std::pair<uint64_t, uint32_t>
+  SetTime(google::protobuf::RepeatedPtrField<valhalla::Location>& locations,
+          const baldr::GraphId& node_id,
+          baldr::GraphReader& reader);
+
+  /**
+   * Expand from the node along the forward search path.
+   * @param graphreader  Graph reader.
+   * @param node Graph Id of the node to expand.
+   * @param pred Edge label of the predecessor edge leading to the node.
+   * @param pred_idx Index in the edge label list of the predecessor edge.
+   * @param from_transition Boolean indicating if this expansion is from a transition edge.
+   * @param localtime Current local time.  Seconds since epoch.
+   * @param seconds_of_week For time dependent Dijkstrass this allows lookup of predicted traffic.
+   */
+  void ExpandForward(baldr::GraphReader& graphreader,
+                     const baldr::GraphId& node,
+                     const sif::EdgeLabel& pred,
+                     const uint32_t pred_idx,
```

Comment:
Whats the purpose of using `const` for primitives that's passed by value? Since its passed by value, it does not affect the callers copy of it, so not sure what we gain by declaring `const` here.

Is it to protect the passed param from accidental change (because this is a long function)?

Reply:
The latter. If you don't intend on changing it, having values immutable by default is a good approach.